README.md
# nd-calendar

[![spm version](http://spmjs.io/badge/nd-calendar)](http://spmjs.io/package/nd-calendar)

> 日历组件

## 安装

```bash
$ spm install nd-calendar --save
```

## 使用

```js
var Calendar = require('nd-calendar');
// use Calendar
```
## 开发

### 本地 Web 服务

```bash
grunt
```

浏览器中访问 http://127.0.0.1:8851

### 生成/查看 API 文档

```bash
grunt doc
grunt
```

浏览器中访问 http://127.0.0.1:8851/doc

### 代码检查与单元测试

```bash
grunt test
```

### 发布组件到 SPM 源

```bash
grunt publish
```


HISTORY.md
# History

---

## 1.0.0

`new` It is the first version of nd-calendar.


tests/index-spec.js
'use strict';

var expect = require('expect.js');
var Calendar = require('../index');

/*globals describe, it*/

describe('nd-calendar', function() {

  it('normal usage', function() {

  });

});


src/date-panel.js
/**
 * User: caolvchong@gmail.com
 * Date: 8/21/13
 * Time: 11:13 AM
 */

  'use strict';
var $ = require('jquery');
var DateUtil = require('nd-date');
var Widget = require('nd-widget');

var helper = {
    getMonthDayCount: function(month, d) {
        var ds = 31;
        if(month == 1) {
            ds = DateUtil.isLeap(d.getFullYear()) ? 29 : 28;
        } else if(month == 3 || month == 5 || month == 8 || month == 10) {
            ds = 30;
        }
        return ds;
    }
};

var defaultFormat = 'yyyy-MM-dd';
var DatePanel = Widget.extend({
    attrs: {
        date: new Date(),
        week: ['一', '二', '三', '四', '五', '六', '日'],
        weekStart: 1, // 一周的起始，默认星期天
        showWeek: false, // 是否显示星期
        format: defaultFormat, // 默认返回格式
        disabled: function(date) { // function, 返回false的则不可点击
            return false;
        }
    },
    events: {
        'click [data-role=set-date]': function(e) {
            var node = $(e.target);
            var prev = DateUtil.format(this.get('date'), this.get('format'));
            var date = node.attr('data-val');
            this.set('date', DateUtil.stringToDate(date, this.get('format')));
            this.trigger('select', date, prev, node);
        },
        'click .date-disabled': function(e) {
            var node = $(e.target);
            var date = DateUtil.format(this.get('date'), this.get('format'));
            var val = node.attr('data-val');
            this.trigger('selectDisabled', val, date, node);
        }
    },
    prev: function() {
        var prev = this.get('date');
        var date = DateUtil.distance(this.get('date'), -1);
        this.set('date', date);
        this.trigger('select', date, prev);
        return this;
    },
    next: function() {
        var prev = this.get('date');
        var date = DateUtil.distance(this.get('date'), 1);
        this.set('date', date);
        this.trigger('select', date, prev);
        return this;
    },
    show: function() {
        var date = this.get('date');
        var week = this.get('week');
        var month = date.getMonth();
        var curr = date.getDate();
        var html = [];
        var weekdayArray = [];
        var weekArr = [];
        var i;
        for(i = 0; i < 7; i++) {
            var t = this.get('weekStart') + i - 1;
            t = t >= 7 ? t - 7 : t;
            weekArr[i] = t;
            weekdayArray[i] = this.get('week')[t];
        }
        html.push('<tr class="widget-calendar-date-column">');
        if(this.get('showWeek')) {
            html.push('<th></th>');
        }
        for(i = 0; i < 7; i++) {
            html.push('<th class="widget-calendar-date widget-calendar-date-' + weekArr[i] + '">' + weekdayArray[i] + '</th>')
        }
        html.push('</tr>');

        var firstDay = new Date(date.getFullYear(), month, 1);
        var firstDayWeek = firstDay.getDay();
        var index = $.inArray(firstDayWeek, weekArr);
        index = index === 0 ? 7 : index;
        var startDay = DateUtil.distance(firstDay, -index + 1);
        var lastDay = DateUtil.distance(firstDay, helper.getMonthDayCount(month, date) - 1);
        var lastDayWeek = lastDay.getDay();
        index = $.inArray(lastDayWeek, weekArr);
        var endDay = DateUtil.distance(lastDay, 7 - index);

        var arr = [];
        var weekCount = ((+endDay - startDay) / (3600 * 24 * 1000) + 1) / 7;
        for(i = 0; i < weekCount; i++) {
            var temp = ['<tr class="widget-calendar-date-panel">'];
            for(var j = 0; j < 7; j++) {
                var d = DateUtil.distance(startDay, i * 7 + j); // 日期对象
                var m = d.getMonth(); // 月份
                var s = DateUtil.format(d, this.get('format')); // 日期字符串
                var n = d.getDate(); // 日期

                temp.push('<td data-val="' + s + '" ');
                if(this.get('disabled').call(this, d) === true) {
                    temp.push('class="date-disabled ')
                    temp.push(m === month ? 'curr-month' : m < month ? 'prev-month' : 'next-month');
                } else {
                    temp.push('data-role="set-date"');
                    temp.push(' class="')
                    if(m === month && n === curr) {
                        temp.push('curr-date ');
                    }
                    temp.push(m === month ? 'curr-month' : m < month ? 'prev-month' : 'next-month');
                }
                temp.push('">' + n + '</td>');
            }
            temp.push('</tr>');
            arr[i] = temp.join('');
        }
        arr.unshift(html.join(''));
        arr.unshift('<table class="widget-calendar-date" data-role="date-panel">');
        arr.push('</table>');
        this.element.html(arr.join('')).show();
        return this;
    }
});

module.exports = DatePanel;


src/time-panel.js
/**
 * User: caolvchong@gmail.com
 * Date: 8/21/13
 * Time: 5:05 PM
 */
'use strict';
var $ = require('jquery');
var Widget = require('nd-widget');
var wheel = require('nd-wheel');
var tpl = require('./../tpl/time.handlebars');

var helper = {
    patchZero: function(n) {
        return n < 10 ? '0' + n : n;
    },
    change: function(input, val) {
        var type = input.attr('data-input');
        val = +val;
        val = isNaN(val) ? (+$.trim(input.val()) || 0) : val;

        if(type === 'hour') {
            val = (val + 24) % 24;
            this._change(val, this.get('minute'), this.get('second'));
        } else {
            val = (val + 60) % 60;
            if(type === 'minute') {
                this._change(this.get('hour'), val, this.get('second'));
            } else {
                this._change(this.get('hour'), this.get('minute'), val);
            }
        }
        input.val(helper.patchZero(val));
    }
};

var TimePanel = Widget.extend({
    attrs: {
        hour: new Date().getHours(),
        minute: new Date().getMinutes(),
        second: new Date().getSeconds(),
        display: {
            hour: true,
            minute: true,
            second: false
        }
    },
    events: {
        'click [data-action]': function(e) {
            var node = $(e.target);
            var isPrev = /prev/.test(node.attr('data-action'));
            var input = node.parent().find('[data-input]');
            var val = +$.trim(input.val()) || 0;
            val = val + (isPrev ? -1 : 1);
            helper.change.call(this, input, val);
        },
        'blur [data-input]': function(e) {
            var input = $(e.target);
            helper.change.call(this, input);
        }
    },
    show: function() {
        var self = this;
        this.element.html(tpl({
            hour: helper.patchZero(this.get('hour')),
            minute: helper.patchZero(this.get('minute')),
            second: helper.patchZero(this.get('second')),
            display: this.get('display')
        })).show();

        this.$('[data-input]').each(function(i, input) {
            input = $(input);
            wheel(input, function(e, type, prevent) {
                prevent();
                if(type === 'up') {
                    helper.change.call(self, input, self.get(input.attr('data-input')) + 1);
                } else if(type === 'down') {
                    helper.change.call(self, input, self.get(input.attr('data-input')) - 1);
                }
                input.select();
            });

            input.hover(function() {
                this.select();
            }, function() {
                this.blur();
            });
        });
        return this;
    },
    output: function() {
        return helper.patchZero(this.get('hour')) + ':' + helper.patchZero(this.get('minute')) + ':' + helper.patchZero(this.get('second'));
    },
    _change: function(hour, minute, second) {
        var flag = false;
        var h = this.get('hour');
        var m = this.get('minute');
        var s = this.get('second');

        if(h !== hour) {
            this.set('hour', hour);
            flag = true;
        }
        if(m !== minute) {
            this.set('minute', minute);
            flag = true;
        }
        if(s !== second) {
            this.set('second', second);
            flag = true;
        }

        if(flag) {
            this.trigger('change', hour, minute, second, h, m, s);
        }
    }
});

module.exports = TimePanel;



src/year-panel.js
/**
 * User: caolvchong@gmail.com
 * Date: 8/20/13
 * Time: 5:44 PM
 */
'use strict';
var $ = require('jquery');
var Widget = require('nd-widget');
var helper = {
    getDate: function(year) {
        return new Date(+year, 0, 1);
    }
};

var YearPanel = Widget.extend({
    attrs: {
        date: new Date(),
        n: 10, // 每页显示的年数个数
        m: 3, // 每行默认显示的个数
        disabled: function(year) { // function, 返回false的则不可点击
            return false;
        },
        prevPlaceholder: '. . .',
        nextPlaceholder: '. . .'
    },
    events: {
        'click [data-role=set-year]': function(e) {
            var node = $(e.target);
            var prev = this.get('date');
            var now = helper.getDate(node.attr('data-val'));
            this.set('date', now);
            this.trigger('select', now, prev, node);
        },
        'click .year-disabled': function(e) {
            var node = $(e.target);
            var year = this.get('date');
            var val = helper.getDate(node.attr('data-val'));
            this.trigger('selectDisabled', val, year, node);
        },
        'click [data-role=get-prev-page]': function(e) {
            var prev = this.get('date');
            var now = helper.getDate(prev.getFullYear() - this.get('n'));
            this.set('date', now);
            this.show();
            this.trigger('select', now, prev);
        },
        'click [data-role=get-next-page]': function(e) {
            var prev = this.get('date');
            var now = helper.getDate(prev.getFullYear() + this.get('n'));
            this.set('date', now);
            this.show();
            this.trigger('select', now, prev);
        }
    },
    prev: function() {
        var prev = this.get('date');
        var now = helper.getDate(prev.getFullYear() - 1);
        this.set('date', now);
        this.trigger('select', now, prev);
        return this;
    },
    next: function() {
        var prev = this.get('date');
        var now = helper.getDate(prev.getFullYear() + 1);
        this.set('date', now);
        this.trigger('select', now, prev);
        return this;
    },
    show: function() {
        var year = this.get('date').getFullYear();

        var n = this.get('n');
        var m = this.get('m');
        var row = Math.ceil((n + 2) / m);
        var start = year - year % n;
        var list = [];
        for(var i = 0; i < n; i++) {
            list[i] = {
                role: 'set-year',
                value: start + i,
                disabled: this.get('disabled').call(this, helper.getDate(start + i))
            };
        }
        list.unshift({
            role: 'get-prev-page',
            value: this.get('prevPlaceholder')
        });
        list.push({
            role: 'get-next-page',
            value: this.get('nextPlaceholder')
        });

        var arr = [];
        var flag = 0;
        for(i = 0; i < row; i++) {
            var temp = ['<tr class="widget-calendar-year-column">'];
            for(var j = flag, len = Math.min(flag + m, list.length); j < len; j++, flag++) {
                var item = list[flag];
                temp.push('<td ')
                if(item.disabled === true) {
                    temp.push('class="year-disabled"');
                } else {
                    temp.push('data-role="' + item.role + '"');
                    if(item.value === year) {
                        temp.push(' class="year-curr"');
                    }
                }
                temp.push(' data-val="' + item.value + '"');
                temp.push('>', item.value, '</td>');
            }
            temp.push('</tr>');
            arr[i] = temp.join('');
        }
        arr.unshift('<table class="widget-calendar-year" data-role="year-panel">');
        arr.push('</table>');

        this.element.html(arr.join('')).show();
        return this;
    }
});

module.exports = YearPanel;


src/month-panel.js
/**
 * User: caolvchong@gmail.com
 * Date: 8/21/13
 * Time: 10:39 AM
 */
'use strict';
var $ = require('jquery');
var DateUtil = require('nd-date');
var Widget = require('nd-widget');

var helper = {
    patchZero: function(n) {
        return n < 10 ? '0' + n : n;
    }
};

var defaultFormat = 'yyyy-MM';
var MonthPanel = Widget.extend({
    attrs: {
        date: new Date(),
        months: ['一月', '二月', '三月', '四月', '五月', '六月', '七月', '八月', '九月', '十月', '十一月', '十二月'],
        disabled: function(month) { // function, 返回false的则不可点击
            return false;
        }
    },
    events: {
        'click [data-role=set-month]': function(e) {
            var node = $(e.target);
            var prev = this.get('date');
            var now = DateUtil.stringToDate(node.attr('data-val'), defaultFormat);
            this.set('date', now);
            this.trigger('select', now, prev, node);
        },
        'click .month-disabled': function(e) {
            var node = $(e.target);
            var month = this.get('date');
            var val = DateUtil.stringToDate(node.attr('data-val'), defaultFormat);
            this.trigger('selectDisabled', val, month, node);
        }
    },
    prev: function() {
        var prev = this.get('date');
        var now = new Date(prev.getFullYear(), prev.getMonth() - 1, 1);
        this.set('date', now);
        this.trigger('select', now, prev);
        return this;
    },
    next: function() {
        var prev = this.get('date');
        var now = new Date(prev.getFullYear(), prev.getMonth() + 1, 1);
        this.set('date', now);
        this.trigger('select', now, prev);
        return this;
    },
    show: function() {
        var date = this.get('date');
        var year = date.getFullYear();
        var month = date.getMonth();
        var list = this.get('months');
        var arr = [];
        var flag = 0;
        for(var i = 0; i < 4; i++) {
            var temp = ['<tr class="widget-calendar-month-column">'];
            for(var j = flag, len = flag + 3; j < len; j++, flag++) {
                temp.push('<td ')
                if(this.get('disabled').call(this, new Date(year, flag, 1)) === true) {
                    temp.push('class="month-disabled"');
                } else {
                    temp.push('data-role="set-month"');
                    if(flag === month) {
                        temp.push(' class="month-curr"');
                    }
                }
                temp.push(' data-val="' + (year + '-' + helper.patchZero(flag + 1)) + '"');
                temp.push('>', list[flag], '</td>');
            }
            temp.push('</tr>');
            arr[i] = temp.join('');
        }
        arr.unshift('<table class="widget-calendar-month" data-role="month-panel">');
        arr.push('</table>');
        this.element.html(arr.join('')).show();
        return this;
    }
});

module.exports = MonthPanel;


index.js
/**
 * @module: nd-calendar
 * @author: lzhengms <lzhengms@gmail.com> - 2015-03-11 13:21:43
 */

'use strict';


var $ = require('jquery');
var Overlay = require('nd-overlay');
var DateUtil = require('nd-date');
var DatePanel = require('./src/date-panel');
var MonthPanel = require('./src/month-panel');
var YearPanel = require('./src/year-panel');
var TimePanel = require('./src/time-panel');
var tpl = {
  calendar: require('./tpl/calendar.handlebars'),
  bar: require('./tpl/bar.handlebars')
};

var defaultFormat = 'yyyy-MM-dd';
var defalutBar = {
  display: {
    clear: true,
    ok: true
  },
  text: {
    clear: '清除',
    ok: '确定'
  }
};

var helper = {
  /**
   * 获取一个日期
   * @param params
   * @returns {*}
   */
  setDate: function(params) {
    var fn = function(a, b, undef) {
      return a === undef ? b : a;
    };
    var d = this.get('date');
    var result;
    if(params.date) {
      var str = params.date;
      var format = defaultFormat;
      if(this.times) {
        str += ' ' + this.times.output();
        format += ' ' + 'HH:mm:ss';
      }
      result = DateUtil.stringToDate(str, format);
    } else {
      if(this.times) {
        result = new Date(fn(params.year, d.getFullYear()), fn(params.month, d.getMonth()), d.getDate(), fn(params.hour, this.times.get('hour')), fn(params.minute, this.times.get('minute')), fn(params.second, this.times.get('second')));
      } else {
        result = new Date(fn(params.year, d.getFullYear()), fn(params.month, d.getMonth()), d.getDate());
      }
    }
    this.set('date', result);
    return result;
  },
  output: function() {
    var output = $(this.get('output'));
    var value = output.val() || output.text();
    if(value) {
      return DateUtil.stringToDate(value, this.get('format'));
    }
  }
};

var Calendar = Overlay.extend({
  attrs: {
    date: {
      value: '',
      getter: function(val) {
        return val || helper.output.call(this) || new Date();
      },
      setter: function(val) {
        if (!val) {
          val = new Date();
        } else if(typeof val === 'string') {
          val = DateUtil.stringToDate(val, this.get('format')) || val;
        }
        return val;
      }
    },
    trigger: null,
    triggerType: 'click',
    output: {
      value: '',
      getter: function(val) {
        val = val ? val : this.get('trigger');
        return $(val);
      }
    },
    mode: 'dates',
    hideOnSelect: true,
    time: false, // 是否显示时间，true显示时分，细粒度控制则传入一个对象{hour: true, minute: true, second: false},
    format: defaultFormat, // 输出格式
    template: tpl.calendar(),
    align: {

      /*getter: function(val) {
       var trigger = $(this.get('trigger'));
       var parentNode = $(this.get('parentNode'));
       var baseElement;
       var baseXY = [0, 0];
       if(trigger && trigger[0]) {
       baseElement = trigger;
       baseXY = [0, trigger.height() + 10];
       } else if(parentNode && parentNode[0]) {
       baseElement = parentNode;
       }
       return {
       selfXY: [0, 0],
       baseElement: baseElement,
       baseXY: baseXY
       };
       }*/

    },
    // 底部按钮栏，默认不显示。当time设置后会自动开启。bar开启后，hideOnSelect会自动转为false
    bar: {
      display: false, // {clear: true, ok: true} 表示：清除,确定按钮的显示与否
      text: defalutBar.text
    },
    disabled: { // date, month, year
      date: function(date) {
        return false;
      },
      month: function(date) {
        return false;
      },
      year: function(date) {
        return false;
      }
    },
    i18n: {},
    view: 'date' // year: 年选择器；month: 年月选择器；date: 正常的日期选择器
  },
  events: {
    'click [data-role=current-month]': function(e) {
      this.renderContainer(this.get('mode') === 'months' ? this.get('view') + 's' : 'months');
    },
    'click [data-role=current-year]': function(e) {
      this.renderContainer(this.get('mode') === 'years' ? this.get('view') + 's' : 'years');
    },
    'click [data-role=prev-year]': function(e) {
      this.years.prev();
      this.renderContainer(this.get('mode'));
    },
    'click [data-role=next-year]': function(e) {
      this.years.next();
      this.renderContainer(this.get('mode'));
    },
    'click [data-role=prev-month]': function(e) {
      this.months.prev();
      this.renderContainer(this.get('mode'));
    },
    'click [data-role=next-month]': function(e) {
      this.months.next();
      this.renderContainer(this.get('mode'));
    },
    'click [data-role="btn-clear"]': function(e) {
      this.output('');
      this.hide();
    },
    'click [data-role="btn-ok"]': function(e) {
      this.output();
      this.hide();
    }
  },
  setup: function() {
    Calendar.superclass.setup.call(this);
    var self = this;
    var container = this.element.find('[data-role=container]');
    var d = this.get('date');
    var bar = this.get('bar');
    var view = this.get('view');
    var mode = view + 's';

    (function(time) { // 存在时间显示的情况下设置默认格式
      if(time) {
        if(this.get('format') === defaultFormat) {
          if(time === true) {
            this.set('time', {
              hour: true,
              minute: true
            });
            this.set('format', defaultFormat + ' ' + 'HH:mm');
          } else {
            var arr = [];
            time.hour = time.hour !== false;
            time.minute = time.minute !== false;
            time.second = time.second === true;
            time.hour && arr.push('HH');
            time.minute && arr.push('mm');
            time.second && arr.push('ss');
            this.set('format', defaultFormat + ' ' + arr.join(':'));
          }
        }
        bar.display = true;
      }
    }).call(this, this.get('time'));

    (function(bar) { // 底部按钮栏控制，开启底部按钮，则选择日期后隐藏日历功能关闭
      if(bar === true) {
        this.set('bar', defalutBar);
        bar = this.get('bar');
      }
      if(bar.display === true) {
        bar.display = defalutBar.display;
      }
      if(bar.display) {
        this.set('hideOnSelect', false);
        container.after(tpl.bar({
          bar: bar
        }));
      }
    }).call(this, bar);

    (function(view) { // 根据模式自动调整输出格式
      if(this.get('format') === defaultFormat) {
        if(view === 'month') {
          this.set('format', 'yyyy-MM');
        } else if(view === 'year') {
          this.set('format', 'yyyy');
        }
      }
      if(typeof this.get('date') === 'string') { // 由于date的setter先于setup，而那时format还未自动处理，这里需要额外处理
        this.set('date', DateUtil.stringToDate(this.get('date'), this.get('format')));
      }
    }).call(this, view);

    this._setPosition();
    this.enable();

    var disabled = this.get('disabled');

    this.set('mode', mode);

    this.dates = new DatePanel({
      date: d,
      week: this.get('i18n').week,
      weekStart: this.get('weekStart'),
      disabled: function(date) {
        return disabled.date.call(self, date) || disabled.month.call(self, date) || disabled.year.call(self, date);
      },
      parentNode: container
    }).render();
    this.months = new MonthPanel({
      date: d,
      months: this.get('i18n').months,
      disabled: function(date) {
        return disabled.month.call(self, date) || disabled.year.call(self, date);
      },
      parentNode: container
    }).render();
    this.years = new YearPanel({
      date: d,
      prevPlaceholder: this.get('i18n').prevPlaceholder,
      nextPlaceholder: this.get('i18n').nextPlaceholder,
      disabled: function(date) {
        return disabled.year.call(self, date);
      },
      parentNode: container
    }).render();
    if(this.get('time')) {
      this.times = new TimePanel({
        parentNode: container,
        display: this.get('time')
      }).render();
      this.times.on('change', function(hour, minute, second) {
        helper.setDate.call(self, {
          hour: hour,
          minute: minute,
          second: second
        });
      });
    }

    this.dates.on('select', function(now, prev, node) {
      var d = helper.setDate.call(self, {
        date: now
      });
      self.renderPannel();
      if(node) {
        self.renderContainer(mode);
        self.get('hideOnSelect') && mode === 'dates' ? self.output() : self.trigger('selectDate', d);
      }
      return false;
    });
    this.months.on('select', function(now, prev, node) {
      var d = helper.setDate.call(self, {
        year: now.getFullYear(),
        month: now.getMonth()
      });
      self.renderPannel();
      if(node && node.attr('data-role') === 'set-month') {
        self.renderContainer(mode);
        self.get('hideOnSelect') && mode === 'months' ? self.output() : self.trigger('selectMonth', d);
      }
    });
    this.years.on('select', function(now, prev, node) {
      var d = helper.setDate.call(self, {
        year: now.getFullYear()
      });
      self.renderPannel();
      if(node && node.attr('data-role') === 'set-year') {
        self.renderContainer(mode);
        self.get('hideOnSelect') && mode === 'years' ? self.output() : self.trigger('selectYear', d);
      }
    });
  },
  show: function() {
    if(!this.rendered) {
      this.render();
    }
    var value = helper.output.call(this) || this.get('date');
    this.set('date', value);
    this.renderPannel();
    this.renderContainer(this.get('mode'));

    Calendar.superclass.show.call(this);
    this.trigger('show');
    return this;
  },
  renderPannel: function() {
    var monthPannel = this.element.find('[data-role=current-month]');
    var yearPannel = this.element.find('[data-role=current-year]');

    var date = this.get('date');
    var month = date.getMonth();
    var year = date.getFullYear();
    monthPannel.text(this.months.get('months')[month]);
    yearPannel.text(year);
  },
  renderContainer: function(mode) {
    var bar = this.$('[data-role=bar]');
    var d = this.get('date');
    this.set('mode', mode);

    this.dates.element.hide();
    this.months.element.hide();
    this.years.element.hide();
    this.times && this.times.element.hide();
    bar.hide();

    this.dates.set('date', d);
    this.months.set('date', d);
    this.years.set('date', d);

    if(mode === 'dates') {
      this.dates.show();
      this.times && this.times.show();
      bar.show();
    } else if(mode === 'months') {
      this.months.show();
    } else if(mode === 'years') {
      this.years.show();
    }
    return this;
  },
  refresh: function() {
    this.renderPannel();
    this.renderContainer(this.get('mode'));
    return this;
  },

  enable: function() {
    var trigger = $(this.get('trigger'));
    if(trigger && trigger[0]) {
      var self = this;
      var event = this.get('triggerType') + '.calendar';
      trigger.on(event, function(e) {
        self.show();
        e.preventDefault();
      });
      this._blurHide(this.get('trigger'));
    }
  },

  disable: function() {
    var trigger = $(this.get('trigger'));
    if(trigger && trigger[0]) {
      var event = this.get('triggerType') + '.calendar';
      trigger.off(event);
    }
  },

  output: function(val, undef) {
    var output = this.get('output');
    var view = this.get('view');
    if(!output.length) {
      return;
    }
    var tagName = output[0].tagName.toLowerCase();
    val = (val === null || val === undef) ? this.get('date') : val;

    var result = val.getDate ? DateUtil.format(val, this.get('format')) : val;
    output[(tagName === 'input' || tagName === 'textarea') ? 'val' : 'text'](result);
    if(this.get('hideOnSelect')) {
      this.hide();
    }
    this.trigger('select' + view.replace(/[a-z]/, function(s) {
      return s.toUpperCase();
    }), typeof val === 'string' ? DateUtil.stringToDate(val) || '' : val);
  }
});

module.exports = Calendar;
